Duplicati/Server/Runner.cs

```diff
@@ -747,6 +758,9 @@ private static void UpdateMetadata(Duplicati.Server.Serialization.Interface.IBac
                 backup.Metadata["LastBackupFinished"] = Library.Utility.Utility.SerializeDateTime(result.EndTime.ToUniversalTime());
                 backup.Metadata["LastBackupDuration"] = r.Duration.ToString();
 
+                if (r.CompactResults != null)
+                    UpdateMetadata(backup, r.CompactResults);

```

Comment:
Does this result in some metadata (e.g., `LastBackupFinished`) being written twice to the database?  For example, we set `LastBackupFinished` a few lines above.  Then, if `r.CompactResults` is not `null`, we update the metadata.  Once that is done, we continue along here and end up updating the metadata again.

Would it suffice to just set the new metadata (`LastCompactDuration`, etc.) here in this `if` statement?

Reply:
Nope, it is not written to the database twice.  The reason I did it this way was so that also an intentional Compact only operation would update the LastCompact time metadata.

For an intentional Compact operation, ICompactResults type is passed to UpdateMetadata.  But for an automatic compaction, it is the parent type IBackupResults that is passed. (presumably because a Backup operation was the main one kicked off).

I was mimicking some previous code in UpdateMetadata that seemed to use the same approach.

Reply:
Sorry, I'm not that familiar with this part of the code.  I haven't looked at it too closely, but it still seems to me that `SetMetadata` at the end of this method can be invoked twice in the case where `IBackupResults` is provided, and `r.CompactResults` is not `null`.

The call to `UpdateMetadata` in the `IBackendStatsticsReporter` case above calls a different overload that simply adds entries to `backup.Metadata`, but doesn't actually insert the entries into the database.  Was that the behavior you were trying to mimic?

Reply:
Yes I saw the one that calls the different overload, but I thought the second one did not. But now looking more closely I see it also calls the different overload.  So yeah good catch!

Because I need to update stats for either ICompactResults (intentional compact) or IBackupResults.CompactResults != nil (auto compact), I could expand the If test higher up instead of calling the function again. It should have the same end result.

I will go ahead and revise....

Reply:
Ok I have reworked this a bit.

It would have been messy to add another overload to `UpdateMetadata` that takes an `ICompactResults` object.  Doing so would have resulted in an intentional compact operation calling the new overload directly, so the Metadata would not actually be written to the database (unless I copied that code from the main `UpdateMetadata` function).

Instead I added a new `UpdateLastCompactMetadata` function and then call it two different ways from the main `UpdateMetadata` function.  One way is used when an intentional compact operation is run, the other is used when compact happened automatically after a backup.

The end result is the same as it was before, but now `UpdateMetadata` doesn't call itself.

Do you like this better?

Reply:
Looks good to me.  Thanks!